### PR TITLE
JAVA-2113: Generate SetEntity DAO methods

### PR DESCRIPTION
**Pull request by olim7t**
Now that the base injection logic is handled by `EntityHelper`, this ticket is very simple: the DAO just needs to create an instance of the helper (cached in a field), and delegate to its `inject` method.

Sample code to test:
```java
@Entity
class User { ... }

@Dao
public interface UserDao {
  @EntityInjector
  void inject(User user, BoundStatement boundStatement);
}
```
(I'm aware that automatic tests are lacking so far, it's on my todo list)

### CODE REVIEW

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java [hunk 1]
```diff
@@ -50,6 +84,20 @@ protected String getFileName() {
   @Override
   protected JavaFile.Builder getContents() {
 
+    List<PartialClassGenerator> methodGenerators = new ArrayList<>();
```
Comment (olim7t) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java]: I will probably remove `PartialClassGenerator` soon: it's not very flexible, I prefer the pattern where the class generator passes itself to child generators and exposes higher-level hooks, such as `addEntityHelperField()` above.

It's an easy refactoring so it can be done later.

**Comment by tomekl007** [comment]
> (I'm aware that automatic tests are lacking so far, it's on my todo list)

How do we plan to test mapper code? 
1. Do we want to examine the generated code line by line?
2. Or maybe we could use generated code and assert that it works as expected?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java [hunk 2]
```diff
@@ -15,33 +15,67 @@
...
     super(context);
     this.interfaceElement = interfaceElement;
     implementationName = GeneratedNames.daoImplementation(interfaceElement);
   }
 
+  /**
+   * Requests the generation of a field holding the {@link EntityHelper} that was generated for the
+   * given entity class, along with the initialization code in the constructor.
+   *
+   * <p>If this is called multiple times, only a single field will be created.
```
Comment (tomekl007) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:64]: Maybe better to state that only the first call will trigger computation?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java [hunk 3]
```diff
@@ -15,33 +15,67 @@
...
+   *
+   * @return the name of the field.
+   */
+  String addEntityHelperField(TypeElement entityClass) {
+    ClassName helperClass = GeneratedNames.entityHelper(entityClass);
+    return entityHelperFields.computeIfAbsent(
+        helperClass,
+        k -> {
+          String baseName =
+              Introspector.decapitalize(entityClass.getSimpleName().toString()) + "Helper";
```
Comment (tomekl007) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:74]: The name will be generated in a camel-case pattern?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java [hunk 4]
```diff
@@ -50,6 +84,20 @@ protected String getFileName() {
...
 
+    List<PartialClassGenerator> methodGenerators = new ArrayList<>();
+    for (Element child : interfaceElement.getEnclosedElements()) {
+      try {
+        if (child.getKind() == ElementKind.METHOD) {
+          ExecutableElement methodElement = (ExecutableElement) child;
+          if (methodElement.getAnnotation(EntityInjector.class) != null) {
+            methodGenerators.add(new EntityInjectorGenerator(methodElement, this, context));
+          }
+          // TODO handle other annotations
```
Comment (tomekl007) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:95]: Is there a need to handle other annotations here? I see that this class is called `DaoImplementationGenerator` so it should it handle only `@Dao` annotation, right? If we will want to handle other annotations here it will no be longer `DaoGenerator`.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java [hunk 5]
```diff
@@ -86,6 +134,20 @@ protected String getFileName() {
...
+      methodGenerator.addConstructorInstructions(constructorBuilder);
+    }
+
+    for (Map.Entry<ClassName, String> entry : entityHelperFields.entrySet()) {
+      ClassName fieldTypeName = entry.getKey();
+      String fieldName = entry.getValue();
+
+      classBuilder.addField(
+          FieldSpec.builder(fieldTypeName, fieldName, Modifier.PRIVATE, Modifier.FINAL).build());
+      constructorBuilder.addStatement("this.$L = new $T(context)", fieldName, fieldTypeName);
```
Comment (tomekl007) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java]: What will be a type of a `context` argument of the generated constructor?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  public EntityInjectorGenerator(
+      ExecutableElement methodElement,
+      DaoImplementationGenerator daoImplementationGenerator,
+      ProcessorContext context) {
+    this.methodElement = methodElement;
+    this.daoImplementationGenerator = daoImplementationGenerator;
+    // We're expecting a void method, where one parameter is an annotated entity, and the other a
+    // SettableByName or subtype
+    TypeMirror returnType = methodElement.getReturnType();
+    if (returnType.getKind() != TypeKind.VOID || methodElement.getParameters().size() != 2) {
```
Comment (tomekl007) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java]: maybe we can name that condition like `isVoidMethodAndTwoHasParameters` and extract such method?
Also, I don't see a check for:
>one parameter is an annotated entity, and the other a
>    // SettableByName or subtype

Only validation on a number of arguments. I see that it is checked after, maybe we can do a pre-validation or all pre-conditions needed at the entry of this processing?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      context
+          .getMessager()
+          .error(
+              methodElement,
+              "%s methods must have two parameters and a void return type",
+              EntityInjector.class.getSimpleName());
+      throw new SkipGenerationException();
```
Comment (tomekl007) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java]: Don't we want to include a message in the `SkipGenerationException`?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        MethodSpec.methodBuilder(methodElement.getSimpleName().toString())
+            .addAnnotation(Override.class)
+            .addModifiers(Modifier.PUBLIC);
+    for (VariableElement parameterElement : methodElement.getParameters()) {
+      overridingMethodBuilder.addParameter(
+          ClassName.get(parameterElement.asType()), parameterElement.getSimpleName().toString());
+    }
+    // Forward to the base injector in the helper:
+    overridingMethodBuilder.addStatement(
+        "$L.inject($L, $L)", helperFieldName, entityParameterName, targetParameterName);
```
Comment (tomekl007) [inline mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java]: what the `inject` call means? We are using some already existing code?

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java (hunk 3 above)
Comment (GregBestland) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:74]: I see a decapatlize here am I missing something?.. why would this be camel case?

Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:74]: Yes, it lower-cases the first letter.
```
Product => productHelper
SomeClass => someClassHelper
```

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java (hunk 4 above)
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java:95]: The generated class will implement all the methods declared on the interface, so we'll eventually handle all "query" annotations here.

```java
@Dao
public interface UserDao {
  @EntityInject // <--
  void inject(User user, BoundStatement statement);

  @Insert // <--
  void save(User user);

  @Select // <--
  User findById(UUID id);
}
```
The loop iterates on the elements _inside_ `UserDao`.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java (hunk 5 above)
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java]: `MapperContext`. It's in the `mapper-runtime` module.

**Comment by olim7t** [comment]
> How do we plan to test mapper code?

I created a ticket to explore a few ideas: [JAVA-2131](https://datastax-oss.atlassian.net/browse/JAVA-2131)

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java (hunk 6 above)
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java]: As we implement more query methods we can extract common patterns if we see them emerge. Right now I'm not sure if `isVoidMethodAndTwoHasParameters` is going to be useful anywhere else.

The check for `SettableByName` is at line 65:
```java
      if (context.getClassUtils().implementsSettableByName(parameterType)) {
```

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java (hunk 7 above)
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java]: The message is passed to `context.getMessager()` on the previous line. It will materialize as a compiler error. It's a bit more convenient to do it here because we have all the context needed, in particular the `Element`, which determines the line of the error.

`SkipGenerationException` is just a simple way to abort quickly when we know that the generation won't succeed. The convention is that the code that throws it should take care of any message, the catching code will just ignore it and proceed (in that case it's `DaoImplementationGenerator.java:97`). It's a bit hacky, but convenient.

Note that we can never throw exceptions from processor code, errors have to be reported via the messager.

File: mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java (hunk 8 above)
Comment (olim7t) [reply mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/EntityInjectorGenerator.java]: Yeah that's a bit hard to follow, it's code from another generated class.

For each `@Entity`, the processor now generates an implementation of `EntityHelper` (in mapper-runtime), that defines `inject()`.
Upon encountering the entity, this class requests that the generated DAO code create an instance of this helper, and store it in a field (line 94):
```java
String helperFieldName = daoImplementationGenerator.addEntityHelperField(entityElement);
```
Then it generates this line of code that delegates to this helper field.

Here's what the generated code looks like:
```java
public class UserDao_Impl implements UserDao {
  private final MapperContext context;

  private final User_Helper userHelper; // The helper field

  private UserDao_Impl(MapperContext context) {
    this.context = context;
    this.userHelper = new User_Helper(context);
  }

  // This is the part that's generated by EntityInjectorGenerator:
  @Override
  public void inject(User user, BoundStatement boundStatement) {
    userHelper.inject(user, boundStatement); // delegate to the helper
  }

  ... // (init methods, not relevant here)
}
```

**Comment by olim7t** [comment]
I pushed a few changes. The DAO implementation class now creates entity helper instances in its `initAsync` method:

```java
  private final User_Helper userHelper;

  private final Address_Helper addressHelper;

  private UserDao_Impl(MapperContext context, User_Helper userHelper,
      Address_Helper addressHelper) {
    this.context = context;
    this.userHelper = userHelper;
    this.addressHelper = addressHelper;
  }

  public static CompletableFuture<UserDao> initAsync(MapperContext context) {
    User_Helper userHelper = new User_Helper(context);
    Address_Helper addressHelper = new Address_Helper(context);
    UserDao_Impl dao = new UserDao_Impl(context,
        userHelper,
        addressHelper);
    return CompletableFuture.completedFuture(dao);
  }
```
This will make sense when we generate methods that need prepared statements.

**Comment by olim7t** [comment]
I just remembered that our bound statements are immutable 🤦‍♂️ `inject` needs to return the `SettableByName`...
 I'll push the required changes shortly.

**Review by GregBestland — APPROVED** [review]
